File: leetcode/python/881_boats_to_save_people.py
class Solution:
    def numRescueBoats(self, people: list[int], limit: int) -> int:
        people.sort()
        l: int = 0
        r:int = len(people) - 1
        count: int = 0
        while l <= r:
            if l < r and people[r] + people[l] <= limit:
                l += 1
            r -= 1
            count += 1
        return count

s = Solution()
print(s.numRescueBoats(people = [1,2], limit = 3))
print(s.numRescueBoats(people = [3,2,2,1], limit = 3))
print(s.numRescueBoats(people = [3,5,3,4], limit = 5))
print(s.numRescueBoats(people = [1,2,4,5], limit = 6))
print(s.numRescueBoats([2,49,10,7,11,41,47,2,22,6,13,12,33,18,10,26,2,6,50,10], 50))

File: leetcode/rust/src/881_boats_to_save_people.rs
impl Solution {
    pub fn num_rescue_boats(people: Vec<i32>, limit: i32) -> i32 {
        let mut people = people;
        people.sort_unstable();
        let mut i = 0;
        let mut j = people.len() - 1;
        let mut boats = 0;
        while i <= j {
            if people[i] + people[j] <= limit {
                i += 1;
            }
            boats += 1;

            // handle usize underflow
            if j > 0 {
                j -= 1;
            }
            else {
                break;
            }
        }
        boats
    }
}
